### PR TITLE
Notify Service Feature Request

### PR DESCRIPTION
**Pull request by shreya-mishra**
Date: `27th nov`

Developer Name: `Shreya`

----

## Issue Ticket Number

-  (https://github.com/Real-Dev-Squad/website-backend/issues/1684)

## Description

Design doc for Notify Service :
https://docs.google.com/document/d/1qOMcp3_IR2ChCGLhz8D3YIg6bYY-njyirGK5Shi1194/edit?usp=sharing

**Test Coverage**
image
image

**Documentation Updated?**

-   [x] Yes
-   [ ] No

_If your feature adds a new API, then documentation related to the feature like API-Contracts, and Data-models  must be updated._

**Breaking Changes**

-   [ ] Yes
-   [x] No

_If your feature introduces breaking changes or if something is missing, please mention the related issue tickets._

**Development Tested?**

-   [x] Yes
-   [ ] No

_Confirm whether the changes have been tested locally during development._

**Tested in Staging?**

-   [x] Yes
-   [ ] No

_Indicate whether the changes have been tested in the staging environment._

**Under Feature Flag**

-   [x] Yes
-   [ ] No

_Specify if the changes are currently under a feature flag._

**Database Changes**

-   [ ] Yes
-   [x] No

_Indicate whether the changes include modifications to the database._

### Screenshots

Attach any relevant screenshots, such as test coverage reports, before and after images, or other visual aids.

## Test Coverage

Provide information on the test coverage for the changes made in this PR. Include details on the types of tests conducted and their outcomes.

## Additional Notes

Include any additional notes, considerations, or explanations that might be helpful for reviewers.

### CODE REVIEW

**Comment by skv93-coder** [comment]
Please add proper description like this https://github.com/Real-Dev-Squad/website-backend/pull/1738

File: controllers/fcmToken.js [hunk 1]
```diff
@@ -0,0 +1,22 @@
+const { saveFcmToken } = require("../models/fcmToken");
+
+/**
+ * Route used to get the health status of teh server
+ *
+ * @param req {Object} - Express request object
+ * @param res {Object} - Express response object
+ */
+const fcmTokenController = (req, res) => {
+  const { fcmToken } = req.body;
```
Comment (heyrandhir) [inline controllers/fcmToken.js:10]: how will we be sure that user will be passing `fcmToken` in the body ? What if they miss sending `fcmToken` in the request payload.

File: controllers/fcmToken.js [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+
+  // Save the FCM token to your database or perform other necessary actions.
+  // You can associate the token with a user or device, for example.
+
+  const fcmTokenId = saveFcmToken({ userId: req.userData.id, fcmToken });
+
+  res.status(200).send(`FCM token saved: ${fcmTokenId}`);
+};
```
Comment (heyrandhir) [inline controllers/fcmToken.js:18]: what if something goes wrong. don't see a try catch block

File: controllers/fcmToken.js [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+ */
+const fcmTokenController = (req, res) => {
+  const { fcmToken } = req.body;
+
+  // Save the FCM token to your database or perform other necessary actions.
+  // You can associate the token with a user or device, for example.
+
+  const fcmTokenId = saveFcmToken({ userId: req.userData.id, fcmToken });
+
+  res.status(200).send(`FCM token saved: ${fcmTokenId}`);
```
Comment (heyrandhir) [inline controllers/fcmToken.js:17]: what will be content type for the response ?

File: controllers/fcmToken.js [hunk 4]
```diff
@@ -0,0 +1,22 @@
...
+ * @param req {Object} - Express request object
+ * @param res {Object} - Express response object
+ */
+const fcmTokenController = (req, res) => {
+  const { fcmToken } = req.body;
+
+  // Save the FCM token to your database or perform other necessary actions.
+  // You can associate the token with a user or device, for example.
+
+  const fcmTokenId = saveFcmToken({ userId: req.userData.id, fcmToken });
```
Comment (heyrandhir) [inline controllers/fcmToken.js:15]: Sending sensitive information like authentication tokens in a response, especially in a client-facing response like a POST response, is generally not recommended. It poses security risks, as the token might be exposed to unintended parties or stored insecurely on the client side.

File: controllers/notify.js
Comment (heyrandhir) [inline controllers/notify.js:1]: The database interaction should be handled within the model file, but it is currently implemented in the controller file. Can you please separate it

File: models/fcmToken.js [hunk 5]
```diff
@@ -0,0 +1,37 @@
+const firestore = require("../utils/firestore");
+
+const fcmTokenModel = firestore.collection("fcmToken");
+
+// eslint-disable-next-line consistent-return
```
Comment (heyrandhir) [inline models/fcmToken.js:5]: 👀 Why haven't we adopted consistent-return

File: models/fcmToken.js [hunk 6]
```diff
@@ -0,0 +1,37 @@
...
+      });
+      return fcmToken.id;
+    } else {
+      let fcmTokenObj = {};
+      fcmTokenSnapshot.forEach((fcmToken) => {
+        fcmTokenObj = {
+          id: fcmToken.id,
+          ...fcmToken.data(),
+        };
+      });
```
Comment (heyrandhir) [inline models/fcmToken.js:22]: what's the use of this loop ?

**Review by heyrandhir — CHANGES_REQUESTED** [review]

File: controllers/notify.js
Comment (shreya-mishra) [reply controllers/notify.js:1]: I already did that in refactor PR

Comment (heyrandhir) [reply controllers/notify.js:1]: image

don't see it there as well

File: models/fcmToken.js (hunk 5 above)
Comment (heyrandhir) [reply models/fcmToken.js:5]: but I see on line 28 there is a return.

File: models/fcmToken.js (hunk 6 above)
Comment (heyrandhir) [reply models/fcmToken.js:22]: wouldn't the fcmTokenObj be over written each time.

Comment (heyrandhir) [reply models/fcmToken.js:22]: can you show an example how it will look after the loop has completed.

**Review by heyrandhir — CHANGES_REQUESTED** [review]

File: controllers/fcmToken.js (hunk 1 above)
Comment (heyrandhir) [reply controllers/fcmToken.js]: what if someone tries to hit the API with POSTMAN

File: controllers/fcmToken.js (hunk 3 above)
Comment (heyrandhir) [reply controllers/fcmToken.js]: can you paste the response you are receiving on postman with headers. I dont think it is a JSON

File: controllers/fcmToken.js [hunk 7]
```diff
@@ -0,0 +1,26 @@
...
+
+    const fcmTokenId = saveFcmToken({ userId: req.userData.id, fcmToken });
+
+    res.status(200).send(`FCM token saved: ${fcmTokenId}`);
+  } catch (error) {
+    throw new Error(error);
+  }
+
+  // Save the FCM token to your database or perform other necessary actions.
+  // You can associate the token with a user or device, for example.
```
Comment (heyrandhir) [inline controllers/fcmToken.js:21]: Do we need the comment

File: controllers/fcmToken.js [hunk 8]
```diff
@@ -0,0 +1,26 @@
...
+ */
+const fcmTokenController = (req, res) => {
+  try {
+    const { fcmToken } = req.body;
+
+    const fcmTokenId = saveFcmToken({ userId: req.userData.id, fcmToken });
+
+    res.status(200).send(`FCM token saved: ${fcmTokenId}`);
+  } catch (error) {
+    throw new Error(error);
```
Comment (heyrandhir) [inline controllers/fcmToken.js:17]: you should return the response with the valid status code instead of throwing error to end user

**Review by heyrandhir — CHANGES_REQUESTED** [review]

**Review by heyrandhir — DISMISSED** [review]

File: controllers/fcmToken.js (hunk 1 above)
Comment (shreya-mishra) [reply controllers/fcmToken.js]: already dsicussed!

**Pull request merged**